android/java/org/chromium/chrome/browser/crypto_wallet/activities/BuySendSwapActivity.java

```diff
@@ -368,30 +333,55 @@ private void getSendSwapQuota(boolean calculatePerSellAsset, boolean sendTx) {
         }
     }
 
+    private SwapParams getSwapParams(String from,
+            Pair<Integer, String> fromInfo, Pair<Integer, String> toInfo, String percent,
+            boolean calculatePerSellAsset) {
+        String value = mFromValueText.getText().toString();
+        String valueTo = mToValueText.getText().toString();
+
+        SwapParams swapParams = new SwapParams();
+        swapParams.takerAddress = from;
+        swapParams.sellAmount = Utils.toWei(value, fromInfo.first, !calculatePerSellAsset);
+        swapParams.buyAmount = Utils.toWei(valueTo, toInfo.first, calculatePerSellAsset);
+        if (swapParams.sellAmount.isEmpty() && swapParams.buyAmount.isEmpty()) {
+            mBtnBuySendSwap.setEnabled(false);
+            mBtnBuySendSwap.setText(getString(R.string.swap));
+
+            return null;
+        }
+        swapParams.buyToken = toInfo.second;
+        swapParams.sellToken = fromInfo.second;
+        try {
+            swapParams.slippagePercentage = Double.parseDouble(percent) / 100;
+        } catch (NumberFormatException ex) {
+        }
+        swapParams.gasPrice = "";
+
+        return swapParams;
+    }
+
     private void workWithSwapQuota(boolean success, SwapResponse response, String errorResponse,
             boolean calculatePerSellAsset, boolean sendTx, String from) {
         if (!success) {
-            SwapResponse nullResponse = new SwapResponse();
-            nullResponse.sellAmount = "0";
-            nullResponse.buyAmount = "0";
-            nullResponse.price = "0";
-            updateSwapControls(nullResponse, calculatePerSellAsset, errorResponse);
+            response = new SwapResponse();
+            response.sellAmount = "0";
+            response.buyAmount = "0";
+            response.price = "0";
             if (errorResponse != null) {
                 if (sendTx) {
                     Log.e(TAG, "Swap error: " + errorResponse);
                 }
             }
-            findViewById(R.id.btn_buy_send_swap).setEnabled(true);
-
-            return;
-        }
-        updateSwapControls(response, calculatePerSellAsset, null);
-        if (sendTx) {
-            TxData data = Utils.getTxData("", Utils.toWeiHex(response.gasPrice),
-                    Utils.toWeiHex(response.estimatedGas), response.to,
-                    Utils.toWeiHex(response.value), Utils.hexStrToNumberArray(response.data));
-            sendSwapTransaction(data, from);
+            mBtnBuySendSwap.setEnabled(true);
+        } else {
+            if (sendTx) {
+                TxData data = Utils.getTxData("", Utils.toWeiHex(response.gasPrice),
+                        Utils.toWeiHex(response.estimatedGas), response.to,
+                        Utils.toWeiHex(response.value), Utils.hexStrToNumberArray(response.data));
+                sendSwapTransaction(data, from);
+            }
         }
+        updateSwapControls(response, calculatePerSellAsset, errorResponse);
```

Comment:
will we have `errorResponse == null` as it used to be how we passed it down in success case?
`updateSwapControls(response, calculatePerSellAsset, null);`

Reply:
I'm assuming if `success` than `errorResponse` will be `null`, or is that not the case?

Reply:
I don't remember, could you check? You don't have to do an actual swap, it's just to enter some values and it should call it when giving you a price quota.

Reply:
Just checked, if `success` is `true` then `errorResponse` is `null`, so it shall be fine.